### PR TITLE
FTheoryTools: Add support for F-theory QSMs

### PR DESCRIPTION
**Pull request by HereAround**
Goal: Support the F-Theory QSMs and their known properties.

Note: The change looks excessive since I added a massive json file with cool data. (Even more, this file was not created by me but by @antonydellavecchia , see https://github.com/antonydellavecchia/polytope-csv-parsing/blob/master/test.json).

cc @antonydellavecchia @apturner

### CODE REVIEW

**Comment by HereAround** [comment]
I have just completed the steps that we discussed in gathertown on Tuesday. So @antonydellavecchia , I believe this is ready for you to take over and complete the next step.

(Also rebased, and hopefully just fixed a silly doctest bug that I introduced a few minutes ago...)

**Comment by HereAround** [comment]
@lgoettgens I see failures from `Aqua.jl` in https://github.com/oscar-system/Oscar.jl/actions/runs/8570358046/job/23488124796?pr=3363. If I recall correctly, you mentioned during this week that this a bug elsewhere and is being investigated? It is of course not impossible that this PR causes this, but I would find it surprising. Most of the (excessively large) changes here are from adding files for a data base and read in data from them in one way or another.

**Comment by lgoettgens** [comment]
> @lgoettgens I see failures from `Aqua.jl` in [oscar-system/Oscar.jl/actions/runs/8570358046/job/23488124796?pr=3363](https://github.com/oscar-system/Oscar.jl/actions/runs/8570358046/job/23488124796?pr=3363). If I recall correctly, you mentioned during this week that this a bug elsewhere and is being investigated? It is of course not impossible that this PR causes this, but I would find it surprising. Most of the (excessively large) changes here are from adding files for a data base and read in data from them in one way or another.

No, that's a different issues that I introduced in a new AA release about an hour ago. I'll make a fix asap.

**Comment by HereAround** [comment]
@antonydellavecchia
* The error "encountered a non-effective model section" was a genuine bug. I am providing the fix in https://github.com/oscar-system/Oscar.jl/pull/3668.
* There were more issues that prevented a loop over all QSM models. In all those cases, this traced back to ill-formatted data in QSMDatabase.json. I have fixed those formatting issues in said .json file. I am not sure if for instance QSMDatabase.mrdi needs to be adjusted?

In any case, once the above fix is merged (if you are curious, you could just cherry-pick this change), the following should work:
```julia
indices = [parse(Int, replace(filename, ".mrdi" => "")) for filename in readdir("Models/QSMDB")]
[literature_model(arxiv_id = "1903.00009", model_parameters = Dict("k" => i)) for i in indices]
```
On my laptop, this takes roughly 30 minutes to complete.

**Comment by HereAround** [comment]
I believe this is ready for review.

**Comment by antonydellavecchia** [comment]
So we have managed to get the size down quite a bit, but we are still look at about > 30mb of storage, so we should still be using an artifact

**Comment by lgoettgens** [comment]
Please do that. I cannot even get the diff of this PR loaded on github as it touches too many files

**Comment by HereAround** [comment]
> So we have managed to get the size down quite a bit, but we are still look at about > 30mb of storage, so we should still be using an artifact

Fine with me. However, I am not sure how/what to do for such an artifact. Can you look into this? Maybe best if we discuss on slack?

File: Artifacts.toml [hunk 1]
```diff
@@ -1,16 +1,15 @@
...
     url = "https://github.com/oscar-system/GAP.jl/releases/download/archive-tag/extraperfect.tar.gz"
-    # Created from https://github.com/hulpke/extraperfect using
-    # #!/bin/sh
-    # export SOURCE_DATE_EPOCH=$(git log -1 --pretty=%ct)
-    # rm -rf extraperfect
-    # mkdir extraperfect
-    # gzip -9 -n -c perf27.grp > extraperfect/perf27.grp.gz
-    # gzip -9 -n -c perf33.grp > extraperfect/perf33.grp.gz
-    # gtar --mtime="$(git log -1 --pretty=%cI)" --owner=0 --group=0 --numeric-owner -cvf extraperfect.tar extraperfect
-    # gzip -n -c extraperfect.tar > extraperfect.tar.gz
```
Comment (benlorenz) [inline Artifacts.toml]: Please keep this comment

**Comment by HereAround** [comment]
In testing this PR with @HechtiDerLachs this morning, I found a hopefully minor issue that needs fixing. I am not (yet) sure what caused it, but hopefully this can be fixed in the existing serialization @antonydellavecchia?

Take the following example (but it most likely also happens for all other QSM models, not just the one with `k = 283`):
```julia
qsm_model = literature_model(arxiv_id = "1903.00009", model_parameters = Dict("k" => 283));
H = qsm_model.hs_model
```
Then we find:
```julia
julia> cox_ring(ambient_space(H))
Multivariate polynomial ring in 14 variables over QQ graded by
  x1 -> [1 0 0 0 0 0 0 0 0]
  x2 -> [0 1 0 0 0 0 0 0 0]
  x3 -> [0 0 1 0 0 0 0 0 0]
  x4 -> [0 0 0 1 0 0 0 0 0]
  x5 -> [0 1 0 -1 0 0 0 0 0]
  x6 -> [1 0 0 0 0 0 0 0 0]
  x7 -> [-2 1 1 1 0 0 0 0 0]
  x8 -> [0 0 0 0 1 0 0 0 0]
  x9 -> [0 0 0 0 0 1 0 0 0]
  x10 -> [0 0 0 0 0 0 1 0 0]
  x11 -> [0 0 0 0 0 0 0 1 0]
  x12 -> [0 0 0 0 0 0 0 0 1]
  x13 -> [0 0 0 0 1 1 0 -1 -2]
  x14 -> [0 0 0 0 0 1 1 1 1]
```
However, there are chosen names for the variables in this toric variety, as we can for instance see from the following:
```julia
julia> parent(hypersurface_equation(H))
Multivariate polynomial ring in 14 variables over QQ graded by
  x1 -> [1 0 0 0 0 0 0 0 0]
  x2 -> [0 1 0 0 0 0 0 0 0]
  x3 -> [0 0 1 0 0 0 0 0 0]
  x4 -> [0 0 0 1 0 0 0 0 0]
  x5 -> [0 1 0 -1 0 0 0 0 0]
  x6 -> [1 0 0 0 0 0 0 0 0]
  x7 -> [-2 1 1 1 0 0 0 0 0]
  v -> [0 0 0 0 1 0 0 0 0]
  e3 -> [0 0 0 0 0 1 0 0 0]
  e2 -> [0 0 0 0 0 0 1 0 0]
  u -> [0 0 0 0 0 0 0 1 0]
  e4 -> [0 0 0 0 0 0 0 0 1]
  e1 -> [0 0 0 0 1 1 0 -1 -2]
  w -> [0 0 0 0 0 1 1 1 1]
```
Indeed, those are the variable names that we want, i.e. there should be u, v, w, e1, e2, e3, e4 appearing. Even more, 
the test `parent(hypersurface_equation(H)) == cox_ring(ambient_space(H))` is required to give `true`.

Here is a hack to achieve this:
```julia
h = hypersurface_equation(H)
S = cox_ring(ambient_space(H))
var_names = symbols(parent(h))
S.R.S = var_names
H.hypersurface_equation = Oscar.eval_poly(string(h), S)
```

File: Artifacts.toml [hunk 2]
```diff
@@ -1,3 +1,11 @@
+[QSMDB]
+git-tree-sha1 = "5c82e5ff3aa99f2aa660c9ffb3caa57ee535cc21"
+lazy = true
+
+     [[QSMDB.download]]
+     sha256 = "3f69aae0fca02d74764f709c622388c564e6e144a8b11347dadef5725a65b0a2"
+     url = "https://github.com/oscar-system/Oscar.jl/releases/download/archive-tag-1/qsmdb.tar.gz"
```
Comment (lgoettgens) [inline Artifacts.toml:7]:
```suggestion
     url = "https://github.com/oscar-system/Oscar.jl/releases/download/archive-tag-1/qsmdb-v1.tar.gz"     
```
To make it possible to change this tar at a later time while still keeping the old version installable, it is good practice to add some kind of version number for the artifact. See e.g. https://github.com/thofma/Hecke.jl/blob/0868996ac2585dc60b5d294a93c9ce94eac24065/Artifacts.toml#L6

Background: once this PR is merged, the artifact file under that location should never change again in the future

Comment (antonydellavecchia) [reply Artifacts.toml:7]: but, is this different from the archive-tag-1 ? where here we use the 1 as the version?

Comment (lgoettgens) [reply Artifacts.toml:7]: That's possible as well, if you want to create a new tag for every new version of the db.

File: experimental/FTheoryTools/src/LiteratureModels/constructors.jl [hunk 3]
```diff
@@ -124,6 +218,73 @@ Hypersurface model over a concrete base
 julia> hypersurface_equation_parametrization(h2)
 b*w*v^2 - c0*u^4 - c1*u^3*v - c2*u^2*v^2 - c3*u*v^3 + w^2
 ```
+A yet more special instance of literature model are the F-theory QSMs. Those consist of 708 families of geometries,
+each of which is obtained from triangulations of polytopes in the 3-dimensional Kreuzer-Skarke list. In particular,
+for each of these 708 families, a lot of information is known. Still, those geometries are also somewhat involved.
+Let us demonstrate this on the F-theory QSM based on the 4th polytope in the 3-dimensional Kreuzer-Skarke list.
+We can create and study this model as follows (TODO: currently under development):
+```jldoctest
```
Comment (lgoettgens) [inline experimental/FTheoryTools/src/LiteratureModels/constructors.jl]:
```suggestion
```jldoctest; setup = :(artifact"QSMDB")
```
This change loads the artifact before running the doctest. Thus there will no longer be verbose printing of the download during doctest execution

Comment (antonydellavecchia) [reply experimental/FTheoryTools/src/LiteratureModels/constructors.jl]: Thanks

Comment (HereAround) [reply experimental/FTheoryTools/src/LiteratureModels/constructors.jl]: The doctests are still failing, but now with a different error: `LoadError: UndefVarError: `@artifact_str` not defined`.

This seems related to the above suggestion. Any suggestions @lgoettgens ?

Comment (lgoettgens) [reply experimental/FTheoryTools/src/LiteratureModels/constructors.jl]: I can try something tomorrow

File: experimental/FTheoryTools/src/LiteratureModels/constructors.jl [hunk 4]
```diff
@@ -124,6 +218,73 @@ Hypersurface model over a concrete base
 julia> hypersurface_equation_parametrization(h2)
 b*w*v^2 - c0*u^4 - c1*u^3*v - c2*u^2*v^2 - c3*u*v^3 + w^2
 ```
+A yet more special instance of literature model are the F-theory QSMs. Those consist of 708 families of geometries,
+each of which is obtained from triangulations of polytopes in the 3-dimensional Kreuzer-Skarke list. In particular,
+for each of these 708 families, a lot of information is known. Still, those geometries are also somewhat involved.
+Let us demonstrate this on the F-theory QSM based on the 4th polytope in the 3-dimensional Kreuzer-Skarke list.
+We can create and study this model as follows (TODO: currently under development):
+```jldoctest; setup = :(artifact"QSMDB")
```
Comment (lgoettgens) [inline experimental/FTheoryTools/src/LiteratureModels/constructors.jl]:
```suggestion
```jldoctest; setup = :(Oscar.LazyArtifacts.artifact"QSMDB")
```
This should fix it

File: experimental/FTheoryTools/src/LiteratureModels/constructors.jl (hunk 3 above)
Comment (lgoettgens) [reply experimental/FTheoryTools/src/LiteratureModels/constructors.jl]: Let's see if my new fix in [`8923fb7` (#3363)](https://github.com/oscar-system/Oscar.jl/pull/3363/commits/8923fb73b709d6e401a629955189daa44e8e45f5) works

Comment (HereAround) [reply experimental/FTheoryTools/src/LiteratureModels/constructors.jl]: Thank you!

File: experimental/FTheoryTools/src/LiteratureModels/constructors.jl [hunk 5]
```diff
@@ -124,6 +218,73 @@ Hypersurface model over a concrete base
 julia> hypersurface_equation_parametrization(h2)
 b*w*v^2 - c0*u^4 - c1*u^3*v - c2*u^2*v^2 - c3*u*v^3 + w^2
 ```
+A yet more special instance of literature model are the F-theory QSMs. Those consist of 708 families of geometries,
+each of which is obtained from triangulations of polytopes in the 3-dimensional Kreuzer-Skarke list. In particular,
+for each of these 708 families, a lot of information is known. Still, those geometries are also somewhat involved.
+Let us demonstrate this on the F-theory QSM based on the 4th polytope in the 3-dimensional Kreuzer-Skarke list.
+We can create and study this model as follows (TODO: currently under development):
+```jldoctest; setup = :(Oscar.LazyArtifacts.artifact"QSMDB")
```
Comment (lgoettgens) [inline experimental/FTheoryTools/src/LiteratureModels/constructors.jl]:
```suggestion
```jldoctest; setup = :(Oscar.LazyArtifacts.ensure_artifact_installed("QSMDB", Oscar.LazyArtifacts.find_artifacts_toml(Oscar.oscardir)))
```

**Comment by HereAround** [comment]
This looks promising now in that the doctests complete successfully. Thank you @lgoettgens !

There are few things to be fixed, e.g. the issue mentioned above about the variables in the cox ring. I believe this could be fixed within this PR and the serialization @antonydellavecchia ?

There are a few other changes: the `literature_model` constructor should return the hs_model with attributes, some of which are meta-data in the corresponding .json file, some others are special for the QSMs such as the dual graph etc. Maybe it is best if I address those changes in a separate PR (since this PR has already been sitting here for 3 months and is now almost ready to go)?

**Comment by antonydellavecchia** [comment]
> This looks promising now in that the doctests complete successfully. Thank you @lgoettgens !
> 
> There are few things to be fixed, e.g. the issue mentioned above about the variables in the cox ring. I believe this could be fixed within this PR and the serialization @antonydellavecchia ?
> 
> There are a few other changes: the `literature_model` constructor should return the hs_model with attributes, some of which are meta-data in the corresponding .json file, some others are special for the QSMs such as the dual graph etc. Maybe it is best if I address those changes in a separate PR (since this PR has already been sitting here for 3 months and is now almost ready to go)?

the changes to the variables has already been made you might need to fetch the new artifact if your still receive the wrong ring

**Comment by HereAround** [comment]
> > This looks promising now in that the doctests complete successfully. Thank you @lgoettgens !
> > There are few things to be fixed, e.g. the issue mentioned above about the variables in the cox ring. I believe this could be fixed within this PR and the serialization @antonydellavecchia ?
> > There are a few other changes: the `literature_model` constructor should return the hs_model with attributes, some of which are meta-data in the corresponding .json file, some others are special for the QSMs such as the dual graph etc. Maybe it is best if I address those changes in a separate PR (since this PR has already been sitting here for 3 months and is now almost ready to go)?
> 
> the changes to the variables has already been made you might need to fetch the new artifact if your still receive the wrong ring

Great. So then maybe we can bring this in as a first working version and I fine tune in a separate PR?

**Review by antonydellavecchia — APPROVED** [review]
Make sure to squash merge so we don't get the files in the history

**Comment by HereAround** [comment]
> Make sure to squash merge so we don't get the files in the history

Just done. Thank you!

**Pull request merged**